File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    static ArrayList<String> answer=new ArrayList<>();
    static int n;
    public String[] solution(String[][] tickets) {
        n=tickets.length;
        boolean[] visited=new boolean[n];
        
        DFS("ICN",tickets,visited,"ICN");
        Collections.sort(answer);
        return answer.get(0).split(" ");
    }
    public static void DFS(String start,String[][] map,boolean[] visited,String word){
        if (word.split(" ").length==n+1){
            answer.add(word);
            return;
        }
        for(int i=0;i<map.length;i++){
            if (map[i][0].equals(start)&&!visited[i]){
                visited[i]=true;
                DFS(map[i][1],map,visited,word+" "+map[i][1]);
                visited[i]=false;
            }
        }
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict
def solution(tickets):
    answer = []
    n=len(tickets)
    my_dict=defaultdict(list)
    for t in tickets:
        my_dict[t[0]].append(t[1]);
        my_dict[t[0]].sort(reverse=True)
    
    stuck=[]
    path=[]
    
    visited=set()
    node_chk=[0]*(n+1)
    
    stuck.append("ICN")
    
    while(len(stuck)>0):
        start=stuck[-1]
        if len(my_dict[start])==0:
            path.append(stuck.pop())
            continue
        end=my_dict[start][-1]
        stuck.append(end)
        my_dict[start]=my_dict[start][:-1]
        
        
    return path[::-1]